server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+	os, err := maybeNewOperatingSystemDB(ctx, ds.writer, hostOS)
+	if err != nil {
+		return err
+	}
+	return maybeUpdateHostOperatingSystemDB(ctx, ds.writer, hostID, os.ID)
+}
+
+// maybeNewOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system. If found,
+// it returns the record including the associated ID. If not found, it returns a call
+// to `newOperatingSystemDB`, which inserts a new record and returns the record
+// including the newly associated ID.
+func maybeNewOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	switch os, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return os, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get operating system")
+	}
+}
+
+// `newOperatingSystemDB` inserts a record for the given operating system and
+// returns the record including the newly associated ID.
+func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version) VALUES (?, ?, ?, ?)"
+	if _, err := tx.ExecContext(ctx, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "insert new operating system")
+	}
+
+	switch storedOS, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return storedOS, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, doRetryErr
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get new operating system")
+	}
+}
+
+// getOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system.
+// If found, it returns the record including the associated ID.
+func getOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	var os fleet.OperatingSystem
+	stmt := "SELECT * FROM operating_systems WHERE name = ? AND version = ? AND arch = ? AND kernel_version = ? LIMIT 1"
+	if err := sqlx.GetContext(ctx, tx, &os, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, err
+	}
+	return &os, nil
+}
+
+// maybeUpdateHostOperatingSystemDB checks the `host_operating_system` table
+// for the given host ID. If found, it compares the given operating system ID
+// against the stored operating system ID and updates the table accordingly.
+// If not found, it inserts a new record for the given host id with the given
+// operating system id.
+func maybeUpdateHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID uint, osID uint) error {
+	switch storedID, err := getIDHostOperatingSystemDB(ctx, tx, hostID); {
+	case storedID == osID:
```

Comment:
switch case states are evaluated top down. You should check errors before checking other returned values.